### PR TITLE
*: fix various data races detected by race detector

### PR DESCRIPTION
**Pull request by xiang90**
I run etcd under high load in real env. No race is detected. 

The races are detected by enabling race detector for integration test. But all of them happens at VERY low frequent (only happens when there is a cluster configuration change)

### CODE REVIEW

File: rafthttp/transport.go [hunk 1]
```diff
@@ -180,7 +180,10 @@ func (t *Transport) Send(msgs []raftpb.Message) {
 		}
 		to := types.ID(m.To)
 
+		t.mu.RLock()
 		p, ok := t.peers[to]
+		t.mu.RUnlock()
```
Comment (yichengq) [inline rafthttp/transport.go]: `mu` is used to `protect the remote and peer map`, and i see `g, ok := t.remotes[to]` 10  lines below, so I think it needs this lock too.

Comment (yichengq) [inline rafthttp/transport.go]: but when running, this cannot happen because we only AddRemote before sending messages.

Comment (xiang90) [inline rafthttp/transport.go]: You can clean this up in the near future. I am trying to clean up the race detected by the race detector in the commit. Or you want me to go over the code?

File: etcdserver/raft.go [hunk 2]
```diff
@@ -52,6 +52,11 @@ const (
 )
 
 var (
+	// protect raftStatus
+	// raftStatus might be accessed concurrently 
+	// by integration tests since multiple etcdservers
+	// share the same raftStatus
+	raftStatusMu sync.Mutex
```
Comment (yichengq) [inline etcdserver/raft.go]: is there any place that raftStatus is called instead of assigned?

Comment (xiang90) [inline etcdserver/raft.go]: This is here only for the integration tests. The raftstatus will only be accessed once. But for safety purpose, we can do more probably.

File: rafthttp/transport.go (hunk 1 above)
Comment (yichengq) [inline rafthttp/transport.go]: This is fine for this PR.

**Comment by yichengq** [comment]
lgtm if CI passes.

**Pull request merged**